src/Perpetuum/Services/Channels/ChatCommands/GameAdminCommands.cs

```diff
@@ -0,0 +1,111 @@
+﻿using Perpetuum.Accounting.Characters;
+using Perpetuum.Host.Requests;
+using Perpetuum.Services.Sessions;
+
+
+namespace Perpetuum.Services.Channels.ChatCommands
+{
+    public class AdminCommandRouter
+    {
+        private readonly GlobalConfiguration _config;
+        private readonly ISessionManager _sessionManager;
+        public AdminCommandRouter(GlobalConfiguration configuration, ISessionManager sessionManager)
+        {
+            _config = configuration;
+            _sessionManager = sessionManager;
+        }
+
+        public void TryParseAdminCommand(Character sender, string text, IRequest request, Channel channel, IChannelManager channelManager)
+        {
+            if (IsAdminCommand(sender, text))
+            {
+                ParseAdminCommand(sender, text, request, channel, channelManager);
+            }
+        }
+
+        private bool IsAdminCommand(Character sender, string message)
+        {
+            return message.Substring(0, 1) == "#" && sender.AccessLevel == AccessLevel.admin;
+        }
+
+        private void ParseAdminCommand(Character sender, string text, IRequest request, Channel channel, IChannelManager channelManager)
+        {
+            if (!IsAdminCommand(sender, text))
+                return;
+
+            string[] command = text.Split(new char[] { ',' });
+
+            var data = AdminCommandData.Create(sender, command, request, channel, channelManager, _sessionManager, _config.EnableDev);
+
+            // channel is not secured. must be secured first.
+            if (channel.Type != ChannelType.Admin)
+            {
+                if (command[0] == "#secure")
+                {
+                    AdminCommandHandlers.Secure(data);
+                    return;
+                }
+                channel.SendMessageToAll(_sessionManager, sender, "Channel must be secured before sending commands.");
+                return;
+            }
+
+            ServerCommands(data);
+        }
+
+        private void ServerCommands(AdminCommandData data)
+        {
+            switch (data.Command.Name)
```

Comment:
What do you think of an attribute that can be put on each handler?
`[GameCommand("shutdown")]`

Reply:
I like it - I'll have to read up on how c# decorators work (attribute?)

Reply:
Ok its not so much like method decorators like in other languages, so maybe I don't understand the purpose.
Is it to use reflection to invoke a command based on the attribute instead of the switch/case here?

Reply:
Well gave it a shot, can't believe it works.  First time implementing something with reflection.